### PR TITLE
fix Issue 19758 - (0x01 & 0xFF) == 0 by dmd 2.085.0(-m64) on Window

### PR DESCRIPTION
**Pull request by rainers**
operation "test SIL,SIL" needs REX prefix

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by rainers** [comment]
Note to reviewers: the wrong code for the condition in the test case was:
```
  48 8D 7D E0        lea         rdi,[rbp-20h]
  4C 8D 04 3A        lea         r8,[rdx+rdi]
  41 0F BE 30        movsx       esi,byte ptr [r8]
  84 F6              test        dh,dh   // wrong register
  75 45              jne         0000000000000121
```
but should have been
```
  48 8D 7D E0        lea         rdi,[rbp-20h]
  4C 8D 04 3A        lea         r8,[rdx+rdi]
  41 0F BE 30        movsx       esi,byte ptr [r8]
  40 84 F6           test        sil,sil    // as expected
  75 45              jne         0000000000000121
```

File: src/dmd/backend/cod3.d [hunk 1]
```diff
@@ -5575,7 +5575,11 @@ void pinholeopt(code *c,block *b)
                 if (op == 0xF7 && rm >= modregrm(3,0,AX) && rm <= modregrm(3,0,7) && (I64 || ereg < 4))
                 {
                     if (u == 0xFF)
+                    {
+                        if (ereg & 4) // SIL,DIL,BPL,SPL need REX prefix
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: I believe this should be `(ereg >= 4)`

**Review by WalterBright — CHANGES_REQUESTED** [review]
see my comment

File: src/dmd/backend/cod3.d (hunk 1 above)
Comment (rainers) [reply src/dmd/backend/cod3.d]: That's what I wrote first, too, but then noticed similar checks using `& 4`, e.g. https://github.com/dlang/dmd/blob/01b7ee6f0906b0fbc3a78b024b0a749d1b56d249/src/dmd/backend/cod3.d#L5989

ereg is limited to < 8 anyway: https://github.com/dlang/dmd/blob/01b7ee6f0906b0fbc3a78b024b0a749d1b56d249/src/dmd/backend/cod3.d#L5360 so it doesn't make a difference.

Comment (WalterBright) [reply src/dmd/backend/cod3.d]: ok

**Pull request merged**